File: travidox_backend/market_data.py
"""
Market Data Provider using Alpha Vantage API
"""

import requests
import os
import time
from typing import Dict, Any, Optional
from datetime import datetime

# Alpha Vantage API endpoint
ALPHA_VANTAGE_URL = "https://www.alphavantage.co/query"

# Cache to store market data and reduce API calls
# Format: {"symbol": {"bid": 1.1234, "ask": 1.1236, "timestamp": 1234567890}}
price_cache = {}
CACHE_EXPIRY = 60  # Cache expiry in seconds

class AlphaVantageProvider:
    """Alpha Vantage API provider for forex data"""
    
    def __init__(self, api_key: Optional[str] = None):
        """Initialize with API key"""
        self.api_key = api_key or os.getenv("ALPHA_VANTAGE_API_KEY", "RDJ0NL7BHOPIA44E")
    
    def get_forex_quote(self, symbol: str) -> Dict[str, Any]:
        """
        Get real-time forex quote for a symbol
        
        Args:
            symbol: Forex symbol (e.g., "EURUSD")
        
        Returns:
            Dict with bid, ask prices and timestamp
        """
        # Check cache first
        now = time.time()
        if symbol in price_cache and (now - price_cache[symbol]["timestamp"]) < CACHE_EXPIRY:
            return price_cache[symbol]
        
        # Parse symbol for Alpha Vantage format (EURUSD -> EUR/USD)
        if len(symbol) == 6 and "/" not in symbol:
            from_currency = symbol[:3]
            to_currency = symbol[3:]
        else:
            parts = symbol.split("/")
            if len(parts) == 2:
                from_currency, to_currency = parts
            else:
                return {
                    "error": f"Invalid symbol format: {symbol}",
                    "bid": None,
                    "ask": None
                }
        
        # Make API request
        params = {
            "function": "CURRENCY_EXCHANGE_RATE",
            "from_currency": from_currency,
            "to_currency": to_currency,
            "apikey": self.api_key
        }
        
        try:
            response = requests.get(ALPHA_VANTAGE_URL, params=params)
            data = response.json()
            
            if "Realtime Currency Exchange Rate" in data:
                exchange_rate = float(data["Realtime Currency Exchange Rate"]["5. Exchange Rate"])
                
                # Alpha Vantage doesn't provide bid/ask directly, so we simulate a spread
                spread = exchange_rate * 0.0002  # 2 pips spread (0.02%)
                bid = exchange_rate - (spread / 2)
                ask = exchange_rate + (spread / 2)
                
                # Format the result
                result = {
                    "bid": round(bid, 5),
                    "ask": round(ask, 5),
                    "timestamp": now,
                    "last_updated": data["Realtime Currency Exchange Rate"]["6. Last Refreshed"]
                }
                
                # Update cache
                price_cache[symbol] = result
                
                return result
            else:
                # Check for error messages
                if "Error Message" in data:
                    return {
                        "error": data["Error Message"],
                        "bid": None,
                        "ask": None
                    }
                elif "Information" in data:
                    return {
                        "error": f"API limit reached: {data['Information']}",
                        "bid": None,
                        "ask": None
                    }
                else:
                    return {
                        "error": "Unknown error from Alpha Vantage API",
                        "bid": None,
                        "ask": None
                    }
                
        except Exception as e:
            return {
                "error": f"Error fetching forex data: {str(e)}",
                "bid": None,
                "ask": None
            }

# Create a singleton instance
_alpha_vantage = None

def get_market_data_provider() -> AlphaVantageProvider:
    """Get or create an Alpha Vantage provider instance"""
    global _alpha_vantage
    if _alpha_vantage is None:
        _alpha_vantage = AlphaVantageProvider()
    return _alpha_vantage 

File: travidox_frontend/lib/market-data.ts
// Market data provider using Finnhub WebSocket API and Alpha Vantage API as fallback

import { getLatestPrice, priceCache as finnhubPriceCache } from './finnhub-websocket';

// Cache to store market data and reduce API calls
// Format: {"symbol": {"bid": 1.1234, "ask": 1.1236, "timestamp": 1234567890}}
const priceCache = new Map<string, MarketPrice>();
const CACHE_EXPIRY = 60 * 1000; // Cache expiry in milliseconds (60 seconds)

export interface MarketPrice {
  bid: number;
  ask: number;
  timestamp: number;
  lastUpdated?: string;
  error?: string;
}

// List of common forex symbols
export const FOREX_SYMBOLS = [
  'EURUSD',
  'GBPUSD',
  'USDJPY',
  'AUDUSD',
  'USDCAD',
  'USDCHF',
  'NZDUSD',
  'EURGBP',
  'EURJPY',
  'GBPJPY'
];

/**
 * Alpha Vantage API provider for forex data
 */
class AlphaVantageProvider {
  private apiKey: string;
  
  constructor(apiKey?: string) {
    // Use provided API key or default
    this.apiKey = apiKey || 'RDJ0NL7BHOPIA44E';
  }
  
  /**
   * Get real-time forex quote for a symbol
   */
  async getForexQuote(symbol: string): Promise<MarketPrice> {
    // Check cache first
    const now = Date.now();
    const cachedPrice = priceCache.get(symbol);
    
    if (cachedPrice && (now - cachedPrice.timestamp) < CACHE_EXPIRY) {
      return cachedPrice;
    }
    
    // Parse symbol for Alpha Vantage format (EURUSD -> EUR/USD)
    let fromCurrency: string;
    let toCurrency: string;
    
    if (symbol.length === 6 && !symbol.includes('/')) {
      fromCurrency = symbol.substring(0, 3);
      toCurrency = symbol.substring(3, 6);
    } else {
      const parts = symbol.split('/');
      if (parts.length === 2) {
        [fromCurrency, toCurrency] = parts;
      } else {
        return {
          bid: 0,
          ask: 0,
          timestamp: now,
          error: `Invalid symbol format: ${symbol}`
        };
      }
    }
    
    // Make API request
    const url = new URL('https://www.alphavantage.co/query');
    url.searchParams.append('function', 'CURRENCY_EXCHANGE_RATE');
    url.searchParams.append('from_currency', fromCurrency);
    url.searchParams.append('to_currency', toCurrency);
    url.searchParams.append('apikey', this.apiKey);
    
    try {
      const response = await fetch(url.toString());
      const data = await response.json();
      
      if (data['Realtime Currency Exchange Rate']) {
        const exchangeRate = parseFloat(data['Realtime Currency Exchange Rate']['5. Exchange Rate']);
        
        // Alpha Vantage doesn't provide bid/ask directly, so we simulate a spread
        const spread = exchangeRate * 0.0002; // 2 pips spread (0.02%)
        const bid = exchangeRate - (spread / 2);
        const ask = exchangeRate + (spread / 2);
        
        // Format the result
        const result: MarketPrice = {
          bid: parseFloat(bid.toFixed(5)),
          ask: parseFloat(ask.toFixed(5)),
          timestamp: now,
          lastUpdated: data['Realtime Currency Exchange Rate']['6. Last Refreshed']
        };
        
        // Update cache
        priceCache.set(symbol, result);
        
        return result;
      } else {
        // Check for error messages
        if (data['Error Message']) {
          return {
            bid: 0,
            ask: 0,
            timestamp: now,
            error: data['Error Message']
          };
        } else if (data['Information']) {
          return {
            bid: 0,
            ask: 0,
            timestamp: now,
            error: `API limit reached: ${data['Information']}`
          };
        } else {
          return {
            bid: 0,
            ask: 0,
            timestamp: now,
            error: 'Unknown error from Alpha Vantage API'
          };
        }
      }
    } catch (error) {
      return {
        bid: 0,
        ask: 0,
        timestamp: now,
        error: `Error fetching forex data: ${error}`
      };
    }
  }
  
  /**
   * Get quotes for multiple symbols
   */
  async getMultipleQuotes(symbols: string[]): Promise<Map<string, MarketPrice>> {
    const results = new Map<string, MarketPrice>();
    
    // Use Promise.all to fetch multiple quotes in parallel
    const promises = symbols.map(symbol => this.getForexQuote(symbol));
    const quotes = await Promise.all(promises);
    
    symbols.forEach((symbol, index) => {
      results.set(symbol, quotes[index]);
    });
    
    return results;
  }
}

/**
 * Simulated market data provider for development/testing
 */
class SimulatedMarketDataProvider {
  private lastPrices: Map<string, MarketPrice>;
  private volatility: number;
  
  constructor(volatility: number = 0.0002) { // 0.02% default volatility
    this.lastPrices = new Map<string, MarketPrice>();
    this.volatility = volatility;
  }
  
  /**
   * Get simulated forex quote
   */
  async getForexQuote(symbol: string): Promise<MarketPrice> {
    const now = Date.now();
    let basePrice: number;
    
    // If we have a last price, use it as the base
    if (this.lastPrices.has(symbol)) {
      basePrice = this.lastPrices.get(symbol)!.bid;
    } else {
      // Generate initial prices based on symbol
      switch (symbol) {
        case 'EURUSD':
          basePrice = 1.0850;
          break;
        case 'GBPUSD':
          basePrice = 1.2650;
          break;
        case 'USDJPY':
          basePrice = 149.50;
          break;
        case 'AUDUSD':
          basePrice = 0.6550;
          break;
        case 'USDCAD':
          basePrice = 1.3650;
          break;
        case 'NZDUSD':
          basePrice = 0.6050;
          break;
        default:
          basePrice = 1.0000;
      }
    }
    
    // Generate random price movement
    const movement = basePrice * this.volatility * (Math.random() * 2 - 1);
    const newBasePrice = basePrice + movement;
    
    // Calculate bid/ask with a spread
    const spread = newBasePrice * 0.0002; // 2 pips spread (0.02%)
    const bid = newBasePrice - (spread / 2);
    const ask = newBasePrice + (spread / 2);
    
    const result: MarketPrice = {
      bid: parseFloat(bid.toFixed(5)),
      ask: parseFloat(ask.toFixed(5)),
      timestamp: now,
      lastUpdated: new Date().toISOString()
    };
    
    // Store the new price
    this.lastPrices.set(symbol, result);
    
    return result;
  }
  
  /**
   * Get quotes for multiple symbols
   */
  async getMultipleQuotes(symbols: string[]): Promise<Map<string, MarketPrice>> {
    const results = new Map<string, MarketPrice>();
    
    // Use Promise.all to fetch multiple quotes in parallel
    const promises = symbols.map(symbol => this.getForexQuote(symbol));
    const quotes = await Promise.all(promises);
    
    symbols.forEach((symbol, index) => {
      results.set(symbol, quotes[index]);
    });
    
    return results;
  }
}

// Create singleton instances
let alphaVantageProvider: AlphaVantageProvider | null = null;
let simulatedProvider: SimulatedMarketDataProvider | null = null;

/**
 * Get market data provider
 */
export function getMarketDataProvider(useSimulated: boolean = false): AlphaVantageProvider | SimulatedMarketDataProvider {
  if (useSimulated) {
    if (!simulatedProvider) {
      simulatedProvider = new SimulatedMarketDataProvider();
    }
    return simulatedProvider;
  } else {
    if (!alphaVantageProvider) {
      alphaVantageProvider = new AlphaVantageProvider();
    }
    return alphaVantageProvider;
  }
}

/**
 * Get market price for a symbol
 */
export async function getMarketPrice(symbol: string, useSimulated: boolean = false): Promise<MarketPrice> {
  // First try to get price from Finnhub WebSocket
  try {
    const finnhubPrice = finnhubPriceCache[symbol];
    if (finnhubPrice && finnhubPrice.bid > 0 && finnhubPrice.ask > 0) {
      return {
        bid: finnhubPrice.bid,
        ask: finnhubPrice.ask,
        timestamp: finnhubPrice.timestamp * 1000, // Convert to milliseconds
        lastUpdated: new Date(finnhubPrice.timestamp * 1000).toISOString()
      };
    }
  } catch (e) {
    console.error('Error getting price from Finnhub:', e);
  }
  
  // If Finnhub data is not available, use Alpha Vantage or simulation
  const provider = getMarketDataProvider(useSimulated);
  return provider.getForexQuote(symbol);
}

/**
 * Get market prices for multiple symbols
 */
export async function getMarketPrices(symbols: string[], useSimulated: boolean = false): Promise<Map<string, MarketPrice>> {
  const results = new Map<string, MarketPrice>();
  
  // First try to get prices from Finnhub WebSocket
  let missingSymbols: string[] = [];
  
  for (const symbol of symbols) {
    try {
      const finnhubPrice = finnhubPriceCache[symbol];
      if (finnhubPrice && finnhubPrice.bid > 0 && finnhubPrice.ask > 0) {
        results.set(symbol, {
          bid: finnhubPrice.bid,
          ask: finnhubPrice.ask,
          timestamp: finnhubPrice.timestamp * 1000, // Convert to milliseconds
          lastUpdated: new Date(finnhubPrice.timestamp * 1000).toISOString()
        });
      } else {
        missingSymbols.push(symbol);
      }
    } catch (e) {
      console.error(`Error getting price for ${symbol} from Finnhub:`, e);
      missingSymbols.push(symbol);
    }
  }
  
  // If any symbols are missing from Finnhub, use Alpha Vantage or simulation
  if (missingSymbols.length > 0) {
    const provider = getMarketDataProvider(useSimulated);
    const fallbackPrices = await provider.getMultipleQuotes(missingSymbols);
    
    for (const [symbol, price] of fallbackPrices.entries()) {
      results.set(symbol, price);
    }
  }
  
  return results;
} 